### DISCUSSION TITLE
Working example?

### DISCUSSION THREAD

**Discussion by theogravity** [hayes/pothos#164] [Q&A] [answered]
Is there a repo that I can look at that integrates a database of some sort to see how a full integration would look like?

**Comment by hayes** [comment]
I don't know of any good open source example apps.  It's something I've been meaning to work on, but haven't had time to build out.

There are functional example apps used for testing in each package.  The only one of these that uses a database is the one for the Prisma plugin: https://github.com/hayes/giraphql/tree/main/packages/plugin-prisma/tests/example

**Reply by theogravity** [reply]
I'm trying to see if I can integrate with an existing GQL application that doesn't use Prisma. The Prisma example doesn't really help me unfortunately.

**Comment by hayes** [comment]
A friend built a scaffolding (paid product) that is designed to be a production ready app that uses GiraphQL.  Probably not what you are looking for though. https://www.nytro.dev/

**Comment by hayes** [comment]
What are the things you would want from ab example app?  Are you looking for something production ready that integrates with a lot of tools (opinionated), something simple that shows off GiraphQL patterns you can adapt into an existing app, etc.   One of the biggest things I've struggled with when trying to start on building out examples is figuring out what should actually be included, how many models, patterns, plugins, 3rd party packages should be part of those examples

**Reply by theogravity** [reply]
Thanks for the responses.

It's really how to use it when a database is involved. Could be something as simple as sqlite.

I'm happy to build an example repo myself, but some guidance would be helpful.

For example, let's say I'm building a simple blog with authors (id, username) and posts (id, title, body, authorId), how would the resolvers look like when using sqlite?

**Reply by theogravity** [reply]
Here's a non-working example (as in I don't have any resolvers that connect to sqlite yet) that I'm currently working on

https://github.com/theogravity/graphql-giraphql-server-client-example/tree/main/server

(I'll prob switch to type-orm from mikro-orm as there's a dataloader for it).

**Reply by theogravity** [reply]
I think this might be the piece I'm missing - does this look correct?

```
builder.queryType({
  fields: (t) => ({
    user: t.field({
      type: User,
      args: {
        id: t.arg.string(),
      },
      resolve: (parent, { id }, context: GQLContext) => {
        // call db to return a user here
      },
    }),
  }),
});
```

Then it would use the def to populate the data?

```
builder.objectType(User, {
  name: 'User',
  description: 'Blog user',
  fields: (t) => ({
    id: t.exposeID('id'),
    name: t.exposeString('name'),
    posts: t.field({
       type: [Post],
       resolve: (parent, null, context: GQLContext) => {
           // db call for posts
       }
    })
  }),
});
```

**Reply by hayes** [reply]
Yup this looks reasonable.  GiraphQL is pretty unopinionated about how you load your data in your resolvers.

General you will connect to your db in the resolver for your query and mutation fields, which will return objects of list of objects (users, posts, etc).  Then in your user and post types you your resolvers can return properties of the parent object (the object loaded by the query,not other parent resolver) without hitting the database.  For fields not loaded in you initial request (like if a users posts) you would define a field with a resolver that does make a db query again and returns the nested object or list of objects.

The exact way you define your types depends on you db.  If you have a class describing your data you can use that as the type param (first arg of builder.objectType) and if you have a type or interface that describes your data, you would use the objectRef API

**Comment by theogravity** [comment, marked as answer]
I've completed my example at https://github.com/theogravity/graphql-giraphql-server-example/

I have some questions regarding typing that I'll create a new discussion topic for. Thanks!

**Reply by hayes** [reply]
Awesome, Nice work!  Sorry I wasn't able to help more. I'm on a road trip with minimal cell service

**Reply by theogravity** [reply]
Thanks for your time, I appreciate it!

**Comment by theogravity** [comment]
@hayes you're free to mention it in documentation if you create an examples section

i'm also happy to transfer it to your repo as well
